File: Discoverer.Logic/GameContract/MarkerSet.cs
using System;
using System.Collections.Immutable;

namespace Discoverer.Logic.GameContract
{
    public record MarkerSet(int? ImproperCellForPlayer, ImmutableHashSet<int> PossibleCellForPlayers)
    {
        public virtual bool Equals(MarkerSet? other)
        {
            if (other is null)
                return false;
            var (improperCellForPlayer, possibleCellForPlayers) = other!;

            return ImproperCellForPlayer == improperCellForPlayer && PossibleCellForPlayers.SetEquals(possibleCellForPlayers);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(ImproperCellForPlayer, PossibleCellForPlayers);
        }
    }
}

File: discoverer.web-view/src/contracts/MarkerSet.ts

class MarkerSet 
{
    improperCellForPlayer: number | null;
    possibleCellForPlayers: Array<number>;

    constructor(
        improperCellForPlayer: number | null,
        possibleCellForPlayers: Array<number>)
    {
        this.improperCellForPlayer = improperCellForPlayer;
        this.possibleCellForPlayers = possibleCellForPlayers;
    }
}

export default MarkerSet;